### PR TITLE
create problem markers from task output

### LINKED ISSUE #4211: support vscode task contribution points
See:
- https://code.visualstudio.com/api/references/contribution-points#contributes.taskDefinitions
- https://code.visualstudio.com/api/references/contribution-points#contributes.problemMatchers
- https://code.visualstudio.com/api/references/contribution-points#contributes.problemPatterns

@marcdumais-work it is required to ship built-in VS Code extensions like `grunt` and `gulp`

### PR DESCRIPTION
**Pull request by elaihau**
What's included in this PR:
- support taskDefinition, problemMatcher, & problemPattern contributed by Theia plugins.
- support customizing tasks by adding named & anonymous problem matchers and patterns in .theia/tasks.json.
- support parsing output from both "start - finish" tasks and "watch" tasks.
- support both single line and multi line patterns.
- create problem markers on parsing task output with the plugin-defined or user-defined problem matchers and patterns.

What's not included:
- prompt users to provide problem matcher and / or patterns before starting tasks
- remove problem markers automatically when problems are fixed

CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19787. Approved on June 4 2019.
CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20139. Approved on June 11 2019.

Resolves #4211

Signed-off-by: elaihau <liang.huang@ericsson.com>

### CODE REVIEW

**Comment by elaihau** [comment]
Quick demo:

1. Support customise tasks by providing named problem matcher(s):

Peek 2019-04-26 14-53


2. Support customise tasks by providing anonymous problem matcher(s) and pattern(s):

Peek 2019-04-26 14-55


3. Support watching tasks and background (aka. watching) pattern(s):

Peek 2019-04-26 15-04

File: packages/markers/src/browser/marker-manager.ts [hunk 1]
```diff
@@ -160,6 +172,22 @@ export abstract class MarkerManager<D extends object> {
         return oldMarkers;
     }
 
+    /*
+     * adds markers for the given uri and owner with the given data, without touching the exsting markers associated with the same uri and owner.
+     */
+    addMarkers(uri: URI, owner: string, data: D[]): Marker<D>[] {
```
Comment (akosyakov) [inline packages/markers/src/browser/marker-manager.ts]: Could we rewrite the client instead that we have a single method `setMarker` which updates them and fire event instead of checking that logic is proper in different methods?

File: packages/markers/src/browser/marker-manager.ts [hunk 2]
```diff
@@ -160,6 +172,22 @@ export abstract class MarkerManager<D extends object> {
...
+        const uriString = uri.toString();
+        const collection = this.uri2MarkerCollection.get(uriString) || new MarkerCollection<D>(uri, this.getKind());
+        const newMarkers = collection.addMarkers(owner, data);
```
Comment (akosyakov) [inline packages/markers/src/browser/marker-manager.ts]: Also code looks strange why adding will delete something? Do you have tests for new APIs?

File: packages/markers/src/browser/marker-manager.ts [hunk 3]
```diff
@@ -197,4 +225,10 @@ export abstract class MarkerManager<D extends object> {
         }
     }
 
+    cleanMarkersByOwner(owner: string): void {
+        this.uri2MarkerCollection.forEach((collection, uri) => {
+            collection.removeMarkers(owner);
+            this.fireOnDidChangeMarkers(new URI(uri));
```
Comment (akosyakov) [inline packages/markers/src/browser/marker-manager.ts]: How do you that event should be fired? Does calling `removeMarkers` removes somethign always? Please add tests.

File: packages/process/src/node/process.ts [hunk 4]
```diff
@@ -108,6 +108,9 @@ export abstract class Process {
         return this.errorEmitter.event;
     }
 
+    abstract onData(listener: (buffer: string) => void): void;
```
Comment (akosyakov) [inline packages/process/src/node/process.ts]: please replace with Theia event API

File: packages/process/src/node/process.ts [hunk 5]
```diff
@@ -108,6 +108,9 @@ export abstract class Process {
...
+    abstract onDataClosed(listener: (exitCode: number, signal?: number) => void): void;
```
Comment (akosyakov) [inline packages/process/src/node/process.ts]: `onClose` or `onEnd` will be better name

Comment (akosyakov) [reply packages/process/src/node/process.ts]: I cannot clients for it. Why do we have it? How is it different from `onExit`?

File: packages/task/src/common/problem-matcher-protocol.ts [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+// This file is inspired by VSCode https://github.com/Microsoft/vscode/blob/1.33.1/src/vs/workbench/contrib/tasks/common/problemMatcher.ts
+// 'problemMatcher.ts' copyright:
+/*---------------------------------------------------------------------------------------------
```
Comment (akosyakov) [inline packages/task/src/common/problem-matcher-protocol.ts]: please open a CQ for all copied code

File: packages/task/src/common/task-watcher.ts [hunk 7]
```diff
@@ -16,31 +16,39 @@
...
         const exitEmitter = this.onTaskExitEmitter;
+        const outputProcessedEmitter = this.onOutputProcessedEmitter;
         return {
             onTaskCreated(event: TaskInfo) {
                 newTaskEmitter.fire(event);
             },
             onTaskExit(event: TaskExitedEvent) {
                 exitEmitter.fire(event);
+            },
+            onTaskOutputProcessed(event: TaskOutputProcessedEvent) {
```
Comment (akosyakov) [inline packages/task/src/common/task-watcher.ts]: `onDidProcessTaskOutput` will be better, we have conventions to event naming similar to VS Code please check, somehow everybody comes with own naming anyway

**Comment by akosyakov** [comment]
@elaihau Does it cover completely problem matching compare to VS Code? How do you verify it? I cannot see any references to some sample VS Code extensions, docs and don't see new tests in the PR.

File: packages/task/src/browser/task-configurations.ts [hunk 8]
```diff
@@ -198,48 +207,46 @@ export class TaskConfigurations implements Disposable {
      * If reading a config file wasn't successful then does nothing.
      */
     protected async refreshTasks(configFileUri: string) {
-        const tasksConfigsArray = await this.readTasks(configFileUri);
-        if (tasksConfigsArray) {
+        const configuredTasksArray = await this.readTasks(configFileUri);
```
Comment (RomanNikitenko) [inline packages/task/src/browser/task-configurations.ts]: Tasks from a config file are read twice per one call refreshTasks() method:
1. `await this.readTasks(configFileUri);` ----> `this.getTaskConfigurationsFromFile(uri);`
2. `await this.readTaskCustomizations(configFileUri);` ----> `this.getTaskConfigurationsFromFile(uri);`

Is it expected?

File: packages/task/src/browser/task-configurations.ts [hunk 9]
```diff
@@ -301,4 +308,50 @@ export class TaskConfigurations implements Disposable {
...
+                    customizations.push(t);
+                }
+            });
+            return customizations;
+        }
+        return undefined;
+    }
+
+    private async isContributedTask(task: TaskConfiguration): Promise<boolean> {
+        const taskDefinition = await this.taskDefinitionRegistry.getDefinition(task);
```
Comment (RomanNikitenko) [inline packages/task/src/browser/task-configurations.ts]: `await` is  unnecessary here

File: packages/task/src/browser/task-configurations.ts [hunk 10]
```diff
@@ -198,48 +207,46 @@ export class TaskConfigurations implements Disposable {
...
-                for (const task of tasksConfigsArray) {
+                for (const task of configuredTasksArray) {
                     newTaskMap.set(task.label, task);
                 }
                 const source = this.getSourceFolderFromConfigUri(configFileUri);
                 this.tasksMap.set(source, newTaskMap);
             }
         }
+
+        const cutomizations = await this.readTaskCustomizations(configFileUri);
```
Comment (RomanNikitenko) [inline packages/task/src/browser/task-configurations.ts]: typo: `cutomizations` ---> cu**s**tomizations

File: packages/markers/src/browser/marker-manager.ts (hunk 1 above)
Comment (elaihau) [reply packages/markers/src/browser/marker-manager.ts]: `setMarkers()` removes existing markers before adding the new ones in.
`addMarkers()` adds new markers to the existing markers collection.
They are doing two different jobs.

I can definitely remove `addMarkers()`, and only use `setMarkers()`. The problem is that, if I have TWO running tasks that happen to have the same owner, markers from one task would overwrite makers from the other. And we will end up with having markers missing in the problem view.

File: packages/markers/src/browser/marker-manager.ts (hunk 2 above)
Comment (elaihau) [reply packages/markers/src/browser/marker-manager.ts]: Yes I shouldn't have the "delete". Will add tests.

File: packages/process/src/node/process.ts (hunk 5 above)
Comment (elaihau) [reply packages/process/src/node/process.ts]: my bad. "onDataClosed" will be removed.

**Comment by elaihau** [comment]
> Does it cover completely problem matching compare to VS Code? 

No. "What's not included" is outlined in the description of the PR. 


> How do you verify it? 

I tested it with running tasks detected by Gulp, Grunt, Jake, and NPM vscode extensions. And to be transparent, Jacques and I found some new missing stuffs when we sat together to test. I am currently working on problems we found.


> I cannot see any references to some sample VS Code extensions, docs 

Could you please be more specific ? I am not sure what you meant by "sample VS Code extensions, docs". 


> and don't see new tests in the PR.

will add tests. I am on it. Thank you !

File: packages/task/src/browser/task-configurations.ts (hunk 8 above)
Comment (elaihau) [reply packages/task/src/browser/task-configurations.ts]: You are right. having one read would be good enough. will improve.

File: packages/markers/src/browser/marker-manager.ts (hunk 1 above)
Comment (akosyakov) [reply packages/markers/src/browser/marker-manager.ts]: no, if you fetch markers for an owner first in taks extension and your markers and then set markers?

Comment (elaihau) [reply packages/markers/src/browser/marker-manager.ts]: ok i reverted changes made to the taskmanager api, and adjusted the client code. also added tests to the classes i added.

File: packages/process/src/node/process.ts (hunk 4 above)
Comment (elaihau) [reply packages/process/src/node/process.ts]: i will take advantage of the `outputStream` Paul added to `abstract class Process` in https://github.com/theia-ide/theia/pull/4586. so i will leave this function as is for now, and rewrite it after PR 4586 gets merged.

Comment (elaihau) [reply packages/process/src/node/process.ts]: ok i removed this function and used `outputStream.on()` instead

File: packages/task/src/common/problem-matcher-protocol.ts (hunk 6 above)
Comment (elaihau) [reply packages/task/src/common/problem-matcher-protocol.ts]: yep, done.

**Comment by elaihau** [comment]
@akosyakov could you please take a look when you get the chance? thank you !

**Comment by elaihau** [comment]
This one has been opened for a while. @akosyakov could you please take a look when you get the chance? thank you !

**Comment by akosyakov** [comment]
@elaihau will review it next month that it is landed in 0.8.0

**Comment by akosyakov** [comment]
@elaihau but you can assign your colleagues and anybody else to help with a review, they should install some vscode extensions and check that everything you fixed is aligned with VS Code

File: packages/task/src/browser/task-definition-registry.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            score += def.properties.all.filter(p => !requiredProps.has(p) && taskConfiguration[p] !== undefined).length;
+            if (score > highest) {
+                highest = score;
+                matchedDefinition = def;
+            }
+        }
+        return matchedDefinition;
+    }
+
+    register(definitionContribution: TaskDefinitionContribution, pluginId: string): void {
```
Comment (akosyakov) [inline packages/task/src/browser/task-definition-registry.ts]: tas extension should not know anything about plugins

Comment (akosyakov) [reply packages/task/src/browser/task-definition-registry.ts]: i think this code should belong to plugin ext, task extension should not know anything about `TaskDefinitionContribution`, but register `TaskDefintion` instead

please review other apis to remove plugin contributions and replace them with internal task extension types

**Comment by akosyakov** [comment]
@elaihau please file follow-up issues for remaining functionality

File: packages/task/src/browser/task-definition-registry.ts (hunk 11 above)
Comment (elaihau) [reply packages/task/src/browser/task-definition-registry.ts]: i did half of the change and get stuck.

Conceptually it makes sense to move plugin contributions to `plugin-ext`, instead of having them in the `task` extension. 
However the plugin contributions don't always come from plugins. they could possibly be defined in the `tasks.json` by users to customise the tasks. and therefore, the `task` extension would need to have the knowledge of transforming `xxxContribution` to `xxx`. 
So, if i move the plugin contributions to the `plugin-ext`, i would need to duplicate some of them in the `task` extension. Otherwise there would be a circular dependency of `task` and `plugin-ext`.

Comment (elaihau) [reply packages/task/src/browser/task-definition-registry.ts]: > task extension should not know anything about plugins

I agree. and like i explained above, some `XxxContribution` interfaces are actually not just for plugins. 
In my opinion, 
- `TaskDefinitionContribution` should renamed into `TaskDefinitionPluginContribution` be moved to `plugin-ext`
- Since `ProblemMatcherContribution`, `WatchingMatcherContribution`, & `ProblemPatternContribution` are for both `plug-ins` & `tasks.json`, they could stay in `task` extension

What do you think? @akosyakov

File: packages/task/src/common/problem-matcher-protocol.ts (hunk 6 above)
Comment (elaihau) [reply packages/task/src/common/problem-matcher-protocol.ts]: CQ has been approved.

**Comment by akosyakov** [comment]
@elaihau I will review it as soon as such SCM damage is undone. I hope we can include it in 0.8.0, but don't want to merge one big refactoring before fixing another.

**Comment by elaihau** [comment]
This PR was tested with following settings:

vsCode built-in matcher: 

- `$eslint-stylish`

- `$go`

- typescript watch

```json
    {
            "type": "shell",
            "label": "build and watch",
            "cwd": "${workspaceFolder}",
            "command": "./node_modules/.bin/tsc -w --preserveWatchOutput",
            "problemMatcher": {
                "name": "tsc-watch",
                "label": "tsc watcher matcher",
                "owner": "typescript",
                "source": "ts",
                "applyTo": "closedDocuments",
                "fileLocation": ["relative", "${cwd}"],
                "pattern":[{
                    "name": "tsc",
                    "regexp": "^([^\\s].*)[\\(:](\\d+)[,:](\\d+)(?:\\):\\s+|\\s+-\\s+)(error|warning|info)\\s+TS(\\d+)\\s*:\\s*(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "code": 5,
                    "message": 6
                }],
                "background": {
                    "activeOnStart": true,
                    "beginsPattern": {
                        "regexp": "^\\s*(?:message TS6032:|\\[?\\d{1,2}:\\d{1,2}:\\d{1,2}(?: AM| PM| a\\.m\\.| p\\.m\\.)?(?:\\]| -)) File change detected\\. Starting incremental compilation\\.\\.\\."
                    },
                    "endsPattern": {
                        "regexp": "^\\s*(?:message TS6042:|\\[?\\d{1,2}:\\d{1,2}:\\d{1,2}(?: AM| PM| a\\.m\\.| p\\.m\\.)?(?:\\]| -)) (?:Compilation complete\\.|Found \\d+ errors?\\.)"
                    }
                }
            }
        }

```

- clang-tidy build task https://clang.llvm.org/extra/clang-tidy/

```json
{
                        "label": "clang-tidy: build",
                        "type": "shell",
                        "cwd": "${workspaceFolder}",
                        "command": "clang-tidy",
                        "args": [
                                "Vector.h"
                        ],
                        "options": {},
                        "problemMatcher": {
                                "name": "clangBuild",
                                "owner": "clang-tidy",
                                "source": "clang-tidy",
                                "applyTo": "alldocuments",
                                "fileLocation": [
                                        "absolute"
                                ],
                                "pattern": [
                                        {
                                                "regexp": "^(.+):(\\d+):(\\d+):\\s+(error|warning|info|note):\\s+(.+?)\\s+\\[(.+)\\]$",
                                                "file": 1,
                                                "line": 2,
                                                "character": 3,
                                                "severity": 4,
                                                "message": 5,
                                                "code": 6
                                        }
                                ]
                        }
                }
```

- a watch task from vsCode repo
```json
		{
			"type": "npm",
			"script": "watch",
			"label": "Build VS Code",
			"group": "build",
			"isBackground": true,
			"presentation": {
				"reveal": "never"
			},
			"problemMatcher": {
				"owner": "typescript",
				"applyTo": "closedDocuments",
				"fileLocation": [
					"absolute"
				],
				"pattern": {
					"regexp": "Error: ([^(]+)\\((\\d+|\\d+,\\d+|\\d+,\\d+,\\d+,\\d+)\\): (.*)$",
					"file": 1,
					"location": 2,
					"message": 3
				},
				"background": {
					"beginsPattern": "Starting compilation",
					"endsPattern": "Finished compilation"
				}
			}
		}
```

File: packages/task/src/browser/task-definition-registry.ts (hunk 11 above)
Comment (akosyakov) [reply packages/task/src/browser/task-definition-registry.ts]: sounds good, although i would get rid of `Contribution` suffix, and just call them `ProblemMatcher`, `WatchingWatcher` and `ProblemPattern`. They can be used as contributions in package.json but they seems to be used like definitions in tasks.json.

File: packages/variable-resolver/src/browser/variable-resolver-service.ts [hunk 12]
```diff
@@ -116,17 +117,25 @@ export namespace VariableResolverService {
     export class Context {
 
         protected readonly resolved = new Map<string, string | undefined>();
+        protected readonly overridden = new Map<string, string | undefined>();
```
Comment (akosyakov) [inline packages/variable-resolver/src/browser/variable-resolver-service.ts]: Why cannot we just use `resolved`?

File: packages/variable-resolver/src/browser/variable-resolver-service.ts [hunk 13]
```diff
@@ -17,11 +17,12 @@
 // tslint:disable:no-any
 
 import { injectable, inject } from 'inversify';
-import { VariableRegistry } from './variable';
+import { Variable, VariableRegistry } from './variable';
 import URI from '@theia/core/lib/common/uri';
 
 export interface VariableResolveOptions {
     context?: URI;
+    overrides?: Variable[];
```
Comment (akosyakov) [inline packages/variable-resolver/src/browser/variable-resolver-service.ts]: It seems that a map of key->value will be enough for now, i.e. `values: { [key: string]: string }`

File: packages/task/src/common/task-protocol.ts [hunk 14]
```diff
@@ -99,9 +109,106 @@ export interface TaskExitedEvent {
...
+
+export interface WatchingMatcherContribution {
+    // If set to true the background monitor is in active mode when the task starts.
+    // This is equals of issuing a line that matches the beginPattern
+    activeOnStart?: boolean;
+    beginsPattern: string | WatchingPattern;
+    endsPattern: string | WatchingPattern;
+}
+
+export interface ProblemMatcherContribution {
```
Comment (akosyakov) [inline packages/task/src/common/task-protocol.ts]: Some properties are never used, please analyse types and create an issue to implement proper support of them, e.g. here `name` and `filePrefix` for example, i did not have a look at other properties and types.

Comment (akosyakov) [reply packages/task/src/common/task-protocol.ts]: or remove used properties and types

File: packages/task/src/browser/task-service.ts [hunk 15]
```diff
@@ -238,18 +290,58 @@ export class TaskService implements TaskConfigurationClient {
...
+                resolvedMatcher = this.problemMatcherRegistry.get(matcher);
+            } else {
+                resolvedMatcher = await this.problemMatcherRegistry.getProblemMatcherFromContribution(matcher);
+            }
+            if (resolvedMatcher) {
+                const scope = task._scope || task._source;
+                if (resolvedMatcher.filePrefix && scope) {
+                    const options = { context: new URI(scope).withScheme('file') };
+                    const cwd = await this.variableResolverService.resolve(task.cwd.toString(), options);
+                    Object.assign(options, {
```
Comment (akosyakov) [inline packages/task/src/browser/task-service.ts]: Should not we register `cwd` variable instead in the variable registry generally? And remove new APIs to the resolver.

It's generally always available in VS Code: https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/workbench/services/configurationResolver/common/variableResolver.ts#L210-L211

**Comment by akosyakov** [comment]
It seems to work somehow. I've noticed that `label` and `cwd` is mandatory. It should not be. `cwd` should be a proper register variable which can be computed from the current context.

Also according to docs, `cwd` should be part of `options`, see here https://code.visualstudio.com/Docs/editor/tasks#_custom-tasks

If the task cannot be started, like `cwd` is missing, there is no any feedback to a user. Plus programming errors in logs.

File: packages/variable-resolver/src/browser/variable-resolver-service.ts (hunk 12 above)
Comment (elaihau) [reply packages/variable-resolver/src/browser/variable-resolver-service.ts]: because variables like ${cwd} could possibly be different from one task to another, e.g.,
```js
[
{ type: 'shell', label: 'task 1', cwd: '/home/user/theia', ... },
{ type: 'shell', label: 'task 2', cwd: '/home/user', ... }
]
```
if i use `resolved` and run 'task 1' and then 'task 2', ${cwd} would be interpreted as '/home/user/theia' for both tasks.

File: packages/task/src/browser/task-service.ts (hunk 15 above)
Comment (elaihau) [reply packages/task/src/browser/task-service.ts]: like i mentioned in another comment,  variables like ${cwd} could possibly be different from one task to another, e.g.,
```js
[
{ type: 'shell', label: 'task 1', cwd: '/home/user/theia', ... },
{ type: 'shell', label: 'task 2', cwd: '/home/user', ... }
]
```

yes, ${cwd} is generally available in vsCode. but i found the variableResolver in vsCode allows users to pass a uri into the function:
https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/workbench/services/configurationResolver/common/variableResolver.ts#L93
and therefore every time ${cwd} gets resolved, the resolver returns the uri passed through. 

do we want to do the same thing?

File: packages/task/src/common/task-protocol.ts (hunk 14 above)
Comment (elaihau) [reply packages/task/src/common/task-protocol.ts]: Thank you for pointing it out. You are right. `filePrefix` is not used. 
`name` is used as the key in the registration.

**Comment by elaihau** [comment]
> It seems to work somehow. I've noticed that `label` and `cwd` is mandatory. It should not be. `cwd` should be a proper register variable which can be computed from the current context.
> 
> Also according to docs, `cwd` should be part of `options`, see here https://code.visualstudio.com/Docs/editor/tasks#_custom-tasks
> 
> If the task cannot be started, like `cwd` is missing, there is no any feedback to a user. Plus programming errors in logs.

You are right. For now the Theia task schema looks a bit different from vsCode's. I am currently working on making them aligned.

Since this PR is already big, can I create a separate github issue for this one and resolve it separately?  @akosyakov

File: packages/task/src/browser/task-definition-registry.ts (hunk 11 above)
Comment (elaihau) [reply packages/task/src/browser/task-definition-registry.ts]: in the most recent push, I updated the interfaces in a little different way.

I realise that the problemMatchers and problemPatterns from tasks.json and package.json differ slightly. 
- developers define problemMatchers and problemPatterns in package.json to **make contributions to the registries, so they always have a name**
- users of tasks.json **associate** problemMatchers and problemPatterns with defined tasks, and therefore the problemMatchers and problemPatterns in tasks.json **don't have to have a name**.

So, in my latest change, `PluginProblemMatcherContribution`, `PluginProblemPatternContribution`, and `PluginTaskDefinitionContribution` are defined in the `plugin-ext` for contributions from plugins. And `ProblemMatcherContribution`, `ProblemPatternContribution`, and `TaskDefinition` are defined in `tasks` for users of tasks.json.

File: packages/task/src/browser/task-service.ts (hunk 15 above)
Comment (akosyakov) [reply packages/task/src/browser/task-service.ts]: Don't we pass context URI from which we can derive a folder URI already?

Comment (elaihau) [reply packages/task/src/browser/task-service.ts]: OK I updated the code to match vscode from the functionality stand point.

**Comment by elaihau** [comment]
Anton, is it possible for you to finish the review before you going on vacation next week? @akosyakov 

I confirmed with @marcdumais-work - if you are concerned about this PR introducing bugs, we could keep it on hold until 0.8 is released. That way we would have a month for bug fixes.

File: packages/plugin-ext/src/plugin/types-impl.ts [hunk 16]
```diff
@@ -1728,12 +1728,14 @@ export class Task {
         if (this.taskExecution instanceof ProcessExecution) {
             Object.assign(this.taskDefinition, {
                 type: 'process',
-                id: this.taskExecution.computeId()
+                id: this.taskExecution.computeId(),
+                taskType: this.taskDefinition!.type
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/plugin/types-impl.ts]: For my own information, are we certain that `taskDefinition` will not be `undefined`?

File: packages/process/src/node/terminal-process.ts [hunk 17]
```diff
@@ -286,5 +286,4 @@ export class TerminalProcess extends Process {
             throw new Error('pty process did not start correctly');
         }
     }
-
```
Comment (vince-fugnitto) [inline packages/process/src/node/terminal-process.ts]: I assume the change is unnecessary?

File: packages/task/src/browser/task-definition-registry.ts [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+import { injectable } from 'inversify';
+import { TaskDefinition, TaskConfiguration, TaskDefinitionRegistry } from '../common';
+
+@injectable()
+export class TaskDefinitionRegistryImpl implements TaskDefinitionRegistry {
+
+    // task type - array of task definitions
+    private definitions: Map<string, TaskDefinition[]> = new Map();
+
+    getDefinitions(taskType: string): TaskDefinition[] {
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-definition-registry.ts]: Can you include documentation for the following method?

File: packages/task/src/browser/task-definition-registry.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        return this.definitions.get(taskType) || [];
+    }
+
+    getDefinition(taskConfiguration: TaskConfiguration): TaskDefinition | undefined {
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-definition-registry.ts]: Include documentation?

File: packages/task/src/browser/task-definition-registry.ts [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+            score += def.properties.all.filter(p => !requiredProps.has(p) && taskConfiguration[p] !== undefined).length;
+            if (score > highest) {
+                highest = score;
+                matchedDefinition = def;
+            }
+        }
+        return matchedDefinition;
+    }
+
+    register(definition: TaskDefinition): void {
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-definition-registry.ts]: Include documentation?

File: packages/task/src/browser/task-definition-registry.ts [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+
+    getDefinitions(taskType: string): TaskDefinition[] {
+        return this.definitions.get(taskType) || [];
+    }
+
+    getDefinition(taskConfiguration: TaskConfiguration): TaskDefinition | undefined {
+        const definitions = this.getDefinitions(taskConfiguration.taskType || taskConfiguration.type);
+        let matchedDefinition: TaskDefinition | undefined;
+        let highest = -1;
+        for (const def of definitions) {
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-definition-registry.ts]: Maybe document what the method attempts to perform?
It's a little hard to follow at the moment, and will only be harder in the future (when looking back).

File: packages/plugin-ext/src/plugin/types-impl.ts (hunk 16 above)
Comment (elaihau) [reply packages/plugin-ext/src/plugin/types-impl.ts]: You are right. resolved as part of https://github.com/theia-ide/theia/pull/5542. Thank you for the review !

**Comment by elaihau** [comment]
i added docs to the public methods. @vince-fugnitto Thank you for the comments !

File: packages/task/src/common/task-protocol.ts [hunk 22]
```diff
@@ -101,9 +111,92 @@ export interface TaskExitedEvent {
...
+    line?: number;
+    character?: number;
+    endLine?: number;
+    endCharacter?: number;
+    code?: number;
+    severity?: number;
+    loop?: boolean;
+}
+
+export const TaskDefinitionRegistry = Symbol('TaskDefinitionRegistry');
```
Comment (akosyakov) [inline packages/task/src/common/task-protocol.ts]: - Why do you need all these symbols and interfaces? Do you have an alternative implementations?
- Why does it belong to protocol? Is it supposed to work over JSON-RPC? 

If not please delete symbols and interrfaces and just use classes intead of them. Otherwise it is just boilerplate. Review code for such things.

Comment (akosyakov) [reply packages/task/src/common/task-protocol.ts]: Added it to the coding guide: https://github.com/theia-ide/theia/wiki/Coding-Guidelines#classes-over-interfaces

File: packages/task/src/common/task-protocol.ts [hunk 23]
```diff
@@ -15,19 +15,23 @@
  ********************************************************************************/
 
 import { JsonRpcServer } from '@theia/core/lib/common/messaging/proxy-factory';
```
Comment (akosyakov) [inline packages/task/src/common/task-protocol.ts]: this file should contain only definition of remote services exposed over JSON-RPC and supporting types

please move everything else to the frontend or backend as classes

**Comment by akosyakov** [comment]
Please update changelog with a record for 0.9.0. If there are breaking changes please add it under breaking changes section.

**Comment by elaihau** [comment]
> Please update changelog with a record for 0.9.0. If there are breaking changes please add it under breaking changes section.

@akosyakov  Thank you for the comments. I updated the changelog. Also I don't see any breaking changes.

File: packages/task/src/common/task-protocol.ts (hunk 22 above)
Comment (elaihau) [reply packages/task/src/common/task-protocol.ts]: done.

File: packages/task/src/node/task.ts [hunk 24]
```diff
@@ -17,19 +17,21 @@
...
 export interface TaskOptions {
-    label: string,
-    config: TaskConfiguration
-    context?: string
+    label: string;
+    config: TaskConfiguration;
+    context?: string;
 }
 
 @injectable()
```
Comment (akosyakov) [inline packages/task/src/node/task.ts]: What is the lifecylce of a task? application-wide? if not then should not it be disposable and be disposed to release listeners and emitters?

Comment (elaihau) [reply packages/task/src/node/task.ts]: what do you think about this change ? https://github.com/theia-ide/theia/pull/5024/commits/fb1e1cbf5e565a8fb43b746499487f7a10427433

`Task` model class now implements Disposable, and will be disposed on `Task.onExit()`
@akosyakov @marechal-p

**Comment by elaihau** [comment]
@benoitf @evidolob I made some changes to the plugin-ext in this PR, are you ok with the change ? Thank you !

**Comment by RomanNikitenko** [comment]
> @RomanNikitenko could you review again ?

I am testing right now, thanks!

**Comment by RomanNikitenko** [comment]
> @RomanNikitenko could you review again ?

@benoitf 
I tested the changes on minikube, started workspace from devfile.
Che tasks defined in devfile (I added simple commands like build and watch for theia project) work as expected except known issues.
The only thing I noticed that tasks from remote plugins are absent in 'Run tasks' menu (in my case I had typescript plugin).
But maybe something wrong with my assembly and typescript plugin has not started.
plugin_manager_error

**Comment by elaihau** [comment]
> The only thing I noticed that tasks from remote plugins are absent in 'Run tasks' menu (in my case I had typescript plugin).

when you said "typescript plugin", did you refer to the `typescript-language-features` from vscode ? That one does not work with Theia. I haven't got the chance to go through its code to figure out what's missing in Theia.

**Comment by elaihau** [comment]
@RomanNikitenko could you please give me more details on the Theia plugin where you see the failure? It would be really nice if you could send me the url, if its source code is available somewhere online, or tell me its name if it comes from VS Code as an extension. Thank you !

**Comment by akosyakov** [comment]
@RomanNikitenko please provide tests with example apps if you can, any contributor should be able to reproduce them without additional setups. If you cannot please let @elaihau know that he can continue  working on compatibility with VS Code

**Comment by RomanNikitenko** [comment]
@elaihau 
I'm sorry for delay, we had three-day weekend and I was offline.
I think you can merge your PR to do not delay the process, because maybe some problem is on our side.
I am testing again and will create an issue if any problem will be found.
Is it OK for you?

**Comment by elaihau** [comment]
@RomanNikitenko no worries ! I will rebase and merge this one. Please file an issue and cc me if you find any problems. Thanks !
